msticpy/vis/heatmap.py

```diff
@@ -0,0 +1,196 @@
+    LayoutDOM
+        The Bokeh plot
+
+    """
+    # Process/extract parameters
+    check_kwargs(kwargs, PlotParams.field_list())
+    param = PlotParams(**kwargs)
+
+    if not param.x_column and not param.y_column:
+        raise ValueError("Must supply `x` and `y` column parameters.")
+
+    reset_output()
+    output_notebook()
+
+    x_range,  y_range = _sort_days_hours(data, param.x_column, param.y_column)
+
+    plot = figure(
+        title=param.title,
+        x_axis_location="above",
+        x_range=x_range,
+        y_range=y_range,
+        plot_width=param.width,
+        plot_height=param.height,
+        tools=["wheel_zoom", "box_zoom", "pan", "reset", "save"],
+        toolbar_location="above",
+    )
+
+    tool_tips = [
+        (param.x_column, f"@{param.x_column} @{param.y_column}:00"),
+        (param.value_col, f"@{param.value_col}")
+    ]
+    plot.add_tools(HoverTool(tooltips=tool_tips))
+
+    mapper, color_bar = _create_colorbar(data, param)
+
+    plot.rect(x=param.y_column, y=param.x_column, width=1, height=1,
+        source=data,
+        fill_color={'field':param.value_col, 'transform': mapper},
+        line_color=None)
+    
+    plot.add_layout(color_bar, 'right')
+
+    _set_plot_params(plot)
+
+    if not param.hide:
+        show(plot)
+    return plot
+
+
+def _set_plot_params(plot):
+    plot.title.text_font_size = "15pt"
+    plot.outline_line_color = None
+    plot.xgrid.visible = True
+    plot.ygrid.visible = True
+    plot.grid.grid_line_color = None
+    plot.grid.grid_line_alpha = 0.1
+    plot.axis.axis_line_color = None
+    plot.axis.major_tick_line_color = None
+    plot.axis.major_label_standoff = 0
+    plot.xaxis.major_label_orientation = pi / 3
+
+def _sort_days_hours(data: pd.DataFrame, week_column: str, hour_column: str):
+    """Sort the Week days and hour of day if required."""
+    dayofweek = list(data[week_column].unique())
+    hourofday = list(data[hour_column].astype(str).unique())
+    correct_days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
+    correct_hours = hours = [f"{hr}" for hr in range(0, 24)]
+    days = {name:val for val, name in enumerate(correct_days)}
+    hours = {name:val for val, name in enumerate(correct_hours)}
+    sorted_days = sorted(dayofweek, key=days.get, reverse=True)
+    sorted_hours = sorted(hourofday, key=hours.get)
+    return  sorted_hours, sorted_days
+
+def _create_colorbar(data: pd.DataFrame, param: PlotParams): 
+    mapper = LinearColorMapper(palette=param.color_pallette, low=data[param.value_col].min(), high=data[param.value_col].max())   
+    color_bar = ColorBar(color_mapper=mapper, major_label_text_font_size=param.major_label_text_font_size,
+                         ticker=BasicTicker(desired_num_ticks=len(param.color_pallette)),
+                         formatter=PrintfTickFormatter(format="%d"),
+                         label_standoff=6, border_line_color=None)
+    return mapper, color_bar
```

Comment:
I'm thinking that we could add a function that will calculate the same data format from any DataFrame.
I thought this would take me a few minutes to knock something up but actually consumed a whole afternoon.

```python

from typing import Optional

X_TIME_DIVS = ["day", "hour"]
Y_TIME_DIVS = ["month", "day"]
y_time_division = "month"
x_time_division = "day"
time_index = pd.Series(pd.date_range(start='2018-01-01', freq='D', periods=7))
day_name_df = (
    pd.DataFrame(time_index.dt.day_name(), columns=["day_name"])
    .reset_index()
)

time_index = pd.Series(pd.date_range(start='2018-01-01', freq='M', periods=12))
month_name_df = (
    pd.DataFrame(time_index.dt.month_name(), columns=["month_name"])
    .reset_index()
)
_hour_index = pd.DataFrame(pd.Series(list(range(24)))).rename(columns={0: "value"})
_day_index = pd.DataFrame(pd.Series(list(range(1, 32)))).rename(columns={0: "value"})


def calculate_heatmap(
    data: pd.DataFrame,
    time_column: str = "TimeGenerated",
    x_time_division: str = "hour",
    y_time_division: str = "day",
    agg_func: str = "count",
    agg_column: Optional[str] = None
) -> pd.DataFrame:
    """
    Return heatmap pivot from DataFrame

    Parameters
    ----------
    data : pd.DataFrame
        Input DataFrame.
    time_column : str, optional
        Name of timestamp column, by default "TimeGenerated"
    x_time_division : str, optional
        Time division to use for X axis, by default "hour"
    y_time_division : str, optional
        Time division to use for Y axis, by default "day"
    agg_func : str, optional
        The aggregate function to apply to the data, by default "count"
    agg_column : Optional[str], optional
        The value column to use to apply the aggregate function.
        This is not needed for the "count" function, it
        defaults to the `time_column` if not supplied explicitly.

    Returns
    -------
    pd.DataFrame
        The pivoted DataFrame to use for Heatmap plot.

    Raises
    ------
    ValueError
        Invalid time division type or x and y time division
        are the same.
    NotImplementedError
        Time division type not yet supported.

    """    
    if x_time_division not in X_TIME_DIVS:
        raise ValueError(f"x_time_division must be one of {', '.join(X_TIME_DIVS)}")
    if y_time_division not in Y_TIME_DIVS:
        raise ValueError(f"y_time_division must be one of {', '.join(Y_TIME_DIVS)}")

    if x_time_division != "hour":
        raise NotImplementedError("day/month chart not yet supported")
    data["x_division"] = data[time_column].dt.hour

    data = (
        data.merge(_hour_index, left_on="x_division", right_on="value", how="right")
        .drop(columns=["x_division"])
        .rename(columns={"value": x_time_division})
    )

    if y_time_division != "day":
        raise NotImplementedError("day/month chart not yet supported")

    if y_time_division == x_time_division:
        raise ValueError(f"x and y cannot be the same time division {y_time_division}")
    data["y_division"] = data[time_column].dt.dayofweek

    y_name_map = day_name_df
    agg_column = agg_column or time_column
    return (
        data[["y_division", x_time_division, agg_column]]
        .groupby(["y_division", x_time_division])
        .agg(agg_func)
        .reset_index()
        .merge(y_name_map, left_on="y_division", right_on="index", how="right")
        .drop(columns=["index", "y_division"])
        .rename(columns={agg_column: "Total"})
    )
```

Reply:
I started to build it so that it would accept month/day as well as day/hour but realized that we couldn't plot that current anyway.